numpy/_core/src/umath/dispatching.cpp

```diff
@@ -981,10 +981,12 @@ promote_and_get_info_and_ufuncimpl_with_locking(
         PyArray_DTypeMeta *op_dtypes[],
         npy_bool legacy_promotion_is_possible)
 {
-    std::shared_mutex *mutex = ((std::shared_mutex *)((PyArrayIdentityHash *)ufunc->_dispatch_cache)->mutex);
-    PyObject *info = PyArrayIdentityHash_GetItemWithLock(
+    NPyRWMutex *mutex = &((PyArrayIdentityHash *)ufunc->_dispatch_cache)->mutex;
```

Comment:
One nice aspect of the separate read lock that you created is that now it is possible to move it inside `PyArrayIdentityHash_GetItem` - that way you do not need to use it here and in `ufunc_object.c`. Furthermore, it guarantees that ordering is always OK at least for this use of your new lock!

(But can do this in follow-up too.)